src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -688,13 +683,62 @@ impl RustcTargetData {
             target_config.insert(ct, host_config.clone());
         }
 
-        Ok(RustcTargetData {
+        let mut res = RustcTargetData {
             rustc,
+            config,
+            requested_kinds: requested_kinds.into(),
             target_config,
             target_info,
             host_config,
             host_info,
-        })
+        };
+
+        // Get all kinds we currently know about. Further kinds will
+        // be added via the `add_kind` function below.
+        //
+        // For now, targets can only ever come from the root workspace
+        // units as artifact dependencies are not a thing yet, so
+        // `add_kind` actually has no caller yet. When we have
+        // artifact dependencies or other ways for targets to appear
+        // at places that are not the root units, we may have to
+        // revisit this.
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| {
+                p.manifest()
+                    .default_kind()
+                    .into_iter()
+                    .chain(p.manifest().forced_kind())
+            }));
+        for kind in all_kinds {
+            res.add_kind(&kind)?;
+        }
+
+        Ok(res)
+    }
+
+    pub fn add_kind(&mut self, kind: &CompileKind) -> CargoResult<()> {
+        if let CompileKind::Target(target) = kind {
+            match self.target_config.entry(*target) {
+                std::collections::hash_map::Entry::Occupied(_) => (),
+                std::collections::hash_map::Entry::Vacant(place) => {
+                    place.insert(self.config.target_cfg_triple(target.short_name())?);
```

Comment:
I think this can probably use `entry(...).or_insert_with(|| ..)` to avoid the `match`?

Reply:
Unfortunately no, `.or_insert_with(|| ...)` would not make it possible to pass the `?` out, as it doesn't make the closure fallible and we need it to be able to call `target.short_name`